### PR TITLE
Cleanup: Eliminate ConstnessAnd

### PR DESCRIPTION
**Pull request by fee1-dead**
This is almost a behaviour-free change and purely a refactoring. "almost" because we appear to be using the wrong ParamEnv somewhere already, and this is now exposed by failing a test using the unstable `~const` feature.

We most definitely need to review all `without_const` and at some point should probably get rid of many of them by using `TraitPredicate` instead of `TraitRef`.

This is a continuation of https://github.com/rust-lang/rust/pull/90274.

r? @oli-obk

cc @spastorino @ecstatic-morse

### CODE REVIEW

**Comment by oli-obk** [comment]
r? @spastorino 

Can't review a PR made up of mostly my own commits

**Comment by oli-obk** [comment]
Tho... essentially @fee1-dead can review this, you only rebased it and fixed some tools. I guess r=oli-obk,fee1-dead if you are alright with my commits

**Comment by fee1-dead** [comment]
I don't think I can do that, as this PR requires some work (fixing the assoc type ui test), also improving caching of queries that do not care about constness if the performance had gotten worse. 

The ui test fix is quite simple. `constness_for_typeck` should return `Constness::Const` when the item is `ImplItem::Fn | ImplItem::Const` in a const impl. I had to remove the function from `rustc_hir` as the crate doesn't know what `TyCtxt` is and therefore cannot query the parent of a node easily.

**Comment by fee1-dead** [comment]
It looks like the try build was successful, but due to me pushing another commit perfbot failed to pick this up, so

@rust-timer build 3bde14d5d8c5c093944947c2dfb90ea047842a0b

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 1]
```diff
@@ -298,9 +298,10 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
 
         let extend = |obligation: traits::PredicateObligation<'tcx>| {
             let mut cause = cause.clone();
-            if let Some(parent_trait_ref) = obligation.predicate.to_opt_poly_trait_ref() {
+            if let Some(parent_trait_ref) = obligation.predicate.to_opt_poly_trait_pred() {
                 let derived_cause = traits::DerivedObligationCause {
-                    parent_trait_ref: parent_trait_ref.value,
+                    // TODO: sus
```
Comment (spastorino) [inline compiler/rustc_trait_selection/src/traits/wf.rs:303]: This is making tidy fail

**Review by spastorino — APPROVED** [review]
The changes in this PR look good to me. Not sure about the perf regression in `deep-vector`.

**Comment by fee1-dead** [comment]
Just modified caching of trait selection. Let's try this once again:

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
> The changes in this PR look good to me. Not sure about the perf regression in deep-vector.

I saw a 4% regression on deep-vector on another perf run while everything else improved by up to 1% today.

**Comment by spastorino** [comment]
I think this now should be good to merge. @fee1-dead if you consider this ready too, feel free to bors=me.

**Comment by spastorino** [comment]
I'm not sure how perf labelling works really but seems like we have `perf-regression` left from the first perf run, but the second shows mixed results.
Should we manually remove `perf-regression` label? if that's the case maybe it's nice if the perf tooling automatically does that in the second run.

cc @rust-lang/wg-compiler-performance

**Comment by fee1-dead** [comment]
> rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

perfbot always marks PRs with regressions as regressions. It automatically removes the label when there are no regressions, which is not the case for this PR.

**Comment by oli-obk** [comment]
I have an idea that could address the remaining regressions: https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/ParamEnv.20and.20constness/near/263477812

**Comment by sdroege** [comment]
This seems to cause a regression, see https://github.com/rust-lang/rust/issues/91489

**Comment by oli-obk** [comment]
Oof. That's quite odd. Let's revert this PR and go with a smaller subset of the changes first?

**Pull request merged**